sql tuning: add EXPLAIN views (#3862)
pingcap/docs: explain-views.md

---

> title: Views

Reviewer: The title is duplicated with the title of https://docs.pingcap.com/tidb/dev/views. What about using another title?

Author: Is there a convention to incorporate path into the title, such as `Explain -> Views`?

Reviewer: `Explain -> Views` does not clearly indicate the content of this document. What about `Views in the Result of EXPLAIN`?

Author: I need to use the same convention with the following content:
```
  Overview
  Explain Walkthough
  Indexes
  Joins
  Subqueries
  Views
  Partitions
```

I will go with `EXPLAIN Statements Using Views`.

Reviewer: `EXPLAIN Statements Using Views` LGTM. Could you make the titles of the list more specific in a similar way?

Author: Yes, I am using the same convention. For example: https://github.com/pingcap/docs/pull/3922

> title: Explain Statements Using Views

Reviewer: ```suggestion
title: EXPALIN Statements Using Views
```

summary: Learn about the execution plan information returned by the `EXPLAIN` statement in TiDB.
---

> # Views

Reviewer: ditto

> 

Reviewer: ```suggestion

This document introduces the execution plan information returned by the `EXPLAIN` statement in TiDB. Several examples are provided.

```

Author: I think this is implied by context. I would prefer if we don't include this to keep the content succinct.

> # Explain Statements Using Views

Reviewer: ```suggestion
# EXPLAIN Statements Using Views
```

> `EXPLAIN` will display the tables and indexes that a [view](/views.md) references, and not the name of the view itself. This is because views are only virtual tables and don't store any data themselves. The definition of the view, and the rest of the statement are merged together during SQL optimization.

Reviewer: ```suggestion
`EXPLAIN` displays the tables and indexes that a [view](/views.md) references, not the name of the view itself. This is because views are only virtual tables and do not store any data themselves. The definition of the view and the rest of the statement are merged together during SQL optimization.
```

> Using the [bikeshare example database](/import-example-data.md), we can see that the following two queries execute in a similar manner:

Reviewer: ```suggestion
From the [bikeshare example database](/import-example-data.md), you can see that the following two queries are executed in a similar manner:
```

+--------------------------------+------------+-----------+---------------------------------------+-------------------------------------+
3 rows in set (0.00 sec)

+-------------------------------+-----------+-----------+---------------------------------------+-------------------------------------+
| id                            | estRows   | task      | access object                         | operator info                       |
+-------------------------------+-----------+-----------+---------------------------------------+-------------------------------------+
| IndexLookUp_10                | 833219.37 | root      |                                       |                                     |
| ├─IndexRangeScan_8(Build)     | 833219.37 | cop[tikv] | table:trips, index:duration(duration) | range:(3600,+inf], keep order:false |
| └─TableRowIDScan_9(Probe)     | 833219.37 | cop[tikv] | table:trips                           | keep order:false                    |
+-------------------------------+-----------+-----------+---------------------------------------+-------------------------------------+
3 rows in set (0.00 sec)
```

> Similarly, predicates from the view will be pushed down to the base table:

Reviewer: ```suggestion
Similarly, predicates from the view are pushed down to the base table:
```

|   └─TableRowIDScan_12          | 3333.33 | cop[tikv] | table:trips                           | keep order:false, stats:pseudo                    |
+--------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------+
4 rows in set (0.00 sec)

+-------------------------+-------------+-----------+---------------+-------------------------------------------+
| id                      | estRows     | task      | access object | operator info                             |
+-------------------------+-------------+-----------+---------------+-------------------------------------------+
| TableReader_7           | 43.00       | root      |               | data:Selection_6                          |
| └─Selection_6           | 43.00       | cop[tikv] |               | eq(bikeshare.trips.bike_number, "W00950") |
|   └─TableFullScan_5     | 19117643.00 | cop[tikv] | table:trips   | keep order:false                          |
+-------------------------+-------------+-----------+---------------+-------------------------------------------+
3 rows in set (0.00 sec)
```

> In the first statement above, we can see that the index was used to satisfy the view definition, and then the `bike_number = 'W00950'` was applied when reading the table row. In the second statement, there were no indexes to satisfy the statement, and a `TableFullScan` was used.

Reviewer: ```suggestion
In the first statement above, you can see that the index is used to satisfy the view definition, and then the `bike_number = 'W00950'` is applied when TiDB reads the table row. In the second statement, there are no indexes to satisfy the statement, and a `TableFullScan` is used.
```

> TiDB will make use of indexes that satisfy both the view definition, and the statement itself. Consider the following composite index:

Reviewer: ```suggestion
TiDB makes use of indexes that satisfy both the view definition and the statement itself. Consider the following composite index:
```

3 rows in set (0.00 sec)

+-------------------------------+----------+-----------+-------------------------------------------------------+---------------------------------------------+
| id                            | estRows  | task      | access object                                         | operator info                               |
+-------------------------------+----------+-----------+-------------------------------------------------------+---------------------------------------------+
| IndexLookUp_10                | 19117.64 | root      |                                                       |                                             |
| ├─IndexRangeScan_8(Build)     | 19117.64 | cop[tikv] | table:trips, index:bike_number(bike_number, duration) | range:["W00950","W00950"], keep order:false |
| └─TableRowIDScan_9(Probe)     | 19117.64 | cop[tikv] | table:trips                                           | keep order:false                            |
+-------------------------------+----------+-----------+-------------------------------------------------------+---------------------------------------------+
3 rows in set (0.00 sec)
```

> In the first statement, TiDB is able to use both parts of the composite index. In the second statement, only the first part of the index was used.

Reviewer 2: ```suggestion
In the first statement, TiDB is able to use both parts of the composite index `(bike_number, duration)`. In the second statement, only the first part `bike_number` of the index `(bike_number, duration)` was used.
```

> In the first statement, TiDB is able to use both parts of the composite index `(bike_number, duration)`. In the second statement, only the first part which is `bike_number` of the index `(bike_number, duration)` was used.

Reviewer: ```suggestion
In the first statement, TiDB is able to use both parts of the composite index `(bike_number, duration)`. In the second statement, only the first part which is `bike_number` of the index `(bike_number, duration)` is used.
```
